sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/hive/rapids/shims/GpuCreateHiveTableAsSelectCommand.scala

```diff
@@ -0,0 +1,116 @@
+                                                  conf: RapidsConf,
+                                                  parent: Option[RapidsMeta[_,_,_]],
+                                                  rule: DataFromReplacementRule)
+  extends DataWritingCommandMeta[CreateHiveTableAsSelectCommand](cmd, conf, parent, rule) {
+
+    var cpuWritingCommand: InsertIntoHiveTable = null
+
+    override def tagSelfForGpu(): Unit = {
+
+      val spark = SparkSession.active
+      val tableDesc = cmd.tableDesc // For the *new* table.
+
+      if (tableDesc.partitionColumnNames.nonEmpty) {
+        willNotWorkOnGpu("partitioned writes are not supported")
+      }
+
+      if (tableDesc.bucketSpec.isDefined) {
+        willNotWorkOnGpu("bucketing is not supported")
+      }
+
+      val catalog = spark.sessionState.catalog
+      val tableExists = catalog.tableExists(tableDesc.identifier)
+      val writingCommand = cmd.getWritingCommand(catalog, tableDesc, tableExists) match {
+        case insertToHiveTable: InsertIntoHiveTable => Some(insertToHiveTable)
+        case c =>
+          willNotWorkOnGpu("Unsupported write command " + c)
+          None
+      }
+
+      if (writingCommand.isDefined) {
+        cpuWritingCommand = writingCommand.get
+        val insertMeta = new GpuInsertIntoHiveTableMeta(
+          cpuWritingCommand,
+          conf,
+          None,
+          GpuOverrides.dataWriteCmds(cpuWritingCommand.getClass))
+
+        insertMeta.tagForGpu()
+        if (!insertMeta.canThisBeReplaced) {
+          insertMeta.getCannotBeReplacedReasons.get.foreach( reason => {
+            willNotWorkOnGpu(reason)
+          })
+        }
+      }
+    }
+
+    override def convertToGpu(): GpuDataWritingCommand = new GpuCreateHiveTableAsSelectCommand(
+      tableDesc = wrapped.tableDesc,
+      query = wrapped.query,
+      outputColumnNames = wrapped.outputColumnNames,
+      mode = wrapped.mode,
+      cpuCmd = wrapped
+    )
+}
+
+case class GpuCreateHiveTableAsSelectCommand(
+    tableDesc: CatalogTable,
+    query: LogicalPlan,
+    outputColumnNames: Seq[String],
+    mode: SaveMode,
+    cpuCmd: CreateHiveTableAsSelectCommand) extends GpuCreateHiveTableAsSelectBase {
+  override def getWritingCommand(
+      catalog: SessionCatalog,
+      tableDesc: CatalogTable,
+      tableExists: Boolean): GpuDataWritingCommand = {
+    // Leverage the existing support for InsertIntoHadoopFsRelationCommand to do the write
+    cpuCmd.getWritingCommand(catalog, tableDesc, tableExists) match {
+      case c: InsertIntoHiveTable =>
+        val rapidsConf = new RapidsConf(conf)
+        val rule = GpuOverrides.dataWriteCmds(c.getClass)
+        val meta = new GpuInsertIntoHiveTableMeta(c, rapidsConf, None, rule)
+        meta.tagForGpu()
+        if (!meta.canThisBeReplaced) {
+          throw new IllegalStateException("Unable to convert writing command: " +
+              meta.explain(all = false))
+        }
+        meta.convertToGpu()
```

Comment:
We are indeed caching it as part of `tagForGpu`. This part is redundant, and should be removed.

Reply:
For more context, this is following the example set in `GpuOptimizedCreateHiveTableAsSelect`. That class should probably be changed as well, because it lands up with a runtime error instead of falling back to CPU.

Reply:
Actually, no. I take it back. Looks like there is a good reason why the writing command cannot be cached ahead of time. It is documented [here](https://github.com/NVIDIA/spark-rapids/blob/branch-23.02/sql-plugin/src/main/340%2B/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala#L177), although I did not completely appreciate it earlier:
```scala
    // It would be cleaner if we could simply call `cmd.getWritingCommand` and let
    // InsertIntoHadoopFsRelationCommandMeta tag the result, but calling getWritingCommand
    // before the table exists will crash. So this ends up replicating a portion of the logic
    // from OptimizedCreateHiveTableAsSelectCommand.getWritingCommand and underlying
    // utility methods to be able to tag whether we can support the optimized Hive write.
```

The `tableDesc` available in `GpuCreateHiveTableAsSelectCommandMeta::tagSelfForGpu()` is the *proposed* table descriptor, prior to table creation. It contains incomplete information (including for table location). Caching the GPU writing command causes the code to crash when it attempts to examine the write's staging directory.

The `tableDesc` presented to `GpuCreateHiveTableAsSelectCommand::getWritingCommand()` is the one *after* the table has been constructed, with full information. This is why the write-command must be reconstructed at runtime.

Reply:
I should document this in the code.

Reply:
> For more context, this is following the example set in GpuOptimizedCreateHiveTableAsSelect. That class should probably be changed as well, because it lands up with a runtime error instead of falling back to CPU.

I stand by this part, though. `GpuOptimizedCreateHiveTableAsSelect` should probably advance its compatibility checks to `tagForGpu`, for fallback.

Reply:
The problem is that this is happening very late, or at least it looks like it is happening after we already moved the CreateHiveTableAsSelect to the GPU.  What happens if we put the GpuCreateHiveTableAsSelect on the GPU but this failed to translate? It looks like we are not going to fall back to the CPU but we are going to cause the job to fail.  That is not a good user experience.  That is what I want to avoid.

Reply:
Agreed. I think I should remove the `tagForGpu()` from `getWritingCommand()`. This is already called during tagging, upstream. We should go straight to `meta.convertToGpu()`.